### PR TITLE
uORB::Publication improvements and cleanup

### PR DESCRIPTION
**Pull request by dagar**
- common PublicationBase
 - PublicationQueued types are now type aliases
 - ORB_PRIO use type everywhere
 - drivers (PX4Accelerometer/PX4Gyroscope/etc) advertise publications in constructor along with driver registration

``` Console
nsh> listener sensor_accel

TOPIC: sensor_accel 3 instances

Instance 0:
 sensor_accel_s
        timestamp: 387779742  (0.004361 seconds ago)
        timestamp_sample: 387779442
        device_id: 3670026 (Type: 0x38, SPI:1 (0x00)) 
        x: 0.1801
        y: -0.3445
        z: -9.8220
        temperature: 39.7980

Instance 1:
 sensor_accel_s
        timestamp: 387781322  (0.004360 seconds ago)
        timestamp_sample: 387781109
        device_id: 3932170 (Type: 0x3C, SPI:1 (0x00)) 
        x: 0.3177
        y: -0.4816
        z: -9.7533
        temperature: 39.3941

Instance 2:
 sensor_accel_s
        timestamp: 387783231  (0.003559 seconds ago)
        timestamp_sample: 387783205
        device_id: 4259850 (Type: 0x41, SPI:1 (0x00)) 
        x: 0.3552
        y: -0.2759
        z: -9.8567
        temperature: 35.0000


nsh> listener sensor_accel_integrated

TOPIC: sensor_accel_integrated 3 instances

Instance 0:
 sensor_accel_integrated_s
        timestamp: 399876375  (0.003065 seconds ago)
        timestamp_sample: 399875924
        error_count: 0
        device_id: 3670026 (Type: 0x38, SPI:1 (0x00)) 
        delta_velocity: [0.0003, -0.0008, -0.0245]
        dt: 2500
        samples: 10
        clip_counter: [0, 0, 0]

Instance 1:
 sensor_accel_integrated_s
        timestamp: 399878665  (0.002044 seconds ago)
        timestamp_sample: 399878352
        error_count: 0
        device_id: 3932170 (Type: 0x3C, SPI:1 (0x00)) 
        delta_velocity: [0.0007, -0.0010, -0.0244]
        dt: 2500
        samples: 10
        clip_counter: [0, 0, 0]

Instance 2:
 sensor_accel_integrated_s
        timestamp: 399882702  (0.000389 seconds ago)
        timestamp_sample: 399882671
        error_count: 0
        device_id: 4259850 (Type: 0x41, SPI:1 (0x00)) 
        delta_velocity: [0.0007, -0.0009, -0.0245]
        dt: 2509
        samples: 2
        clip_counter: [0, 0, 0]



nsh> listener sensor_accel_status

TOPIC: sensor_accel_status 3 instances

Instance 0:
 sensor_accel_status_s
        timestamp: 409639646  (0.100235 seconds ago)
        error_count: 0
        device_id: 3670026 (Type: 0x38, SPI:1 (0x00)) 
        temperature: 39.7980
        clipping: [0, 0, 0]
        full_scale_range: 156.9064
        vibration_metric: 0.0001
        measure_rate_hz: 800
        rotation: 2

Instance 1:
 sensor_accel_status_s
        timestamp: 409643983  (0.097395 seconds ago)
        error_count: 0
        device_id: 3932170 (Type: 0x3C, SPI:1 (0x00)) 
        temperature: 39.3941
        clipping: [0, 0, 0]
        full_scale_range: 156.9064
        vibration_metric: 0.0002
        measure_rate_hz: 800
        rotation: 2

Instance 2:
 sensor_accel_status_s
        timestamp: 409696031  (0.046810 seconds ago)
        error_count: 0
        device_id: 4259850 (Type: 0x41, SPI:1 (0x00)) 
        temperature: 35.0000
        clipping: [0, 0, 0]
        full_scale_range: 156.9064
        vibration_metric: 0.0043
        measure_rate_hz: 1000
        rotation: 10
```

### CODE REVIEW

File: src/modules/sensors/voted_sensors_update.cpp [hunk 1]
```diff
@@ -702,14 +702,16 @@ bool VotedSensorsUpdate::checkFailover(SensorData &sensor, const char *sensor_na
...
-				sensor.priority[failover_index] = 1;
+				sensor.priority[failover_index] = ORB_PRIO_MIN;
 
 				PX4_ERR("Sensor %s #%i failed. Reconfiguring sensor priorities.", sensor_name, failover_index);
 
 				int ctr_valid = 0;
 
 				for (uint8_t i = 0; i < sensor.subscription_count; i++) {
-					if (sensor.priority[i] > 1) { ctr_valid++; }
+					if (sensor.priority[i] > 1) {
```
Comment (bkueng) [inline src/modules/sensors/voted_sensors_update.cpp:712]: You don't want to use `ORB_PRIO_MIN` here?

File: src/modules/uORB/Publication.hpp [hunk 2]
```diff
@@ -40,16 +40,47 @@
...
+
+protected:
+
+	PublicationBase(ORB_ID id) : _orb_id(id) {}
+
+	~PublicationBase()
+	{
+		if (_handle != nullptr) {
+			// don't automatically unadvertise queued publications (eg vehicle_command)
+			if (static_cast<DeviceNode *>(_handle)->get_queue_size() == 1) {
```
Comment (bkueng) [inline src/modules/uORB/Publication.hpp:69]: We need to double-check if that works as expected. No case specifically comes to mind, `ulog_stream_ack` might be a candidate.

**Comment by dagar** [comment]
> Is the ORB_PRIO uint8_t -> int change noticeable?

Overall we save a small amount with this PR (maybe a few hundred bytes), so it seems uORB::Publication using an ORB_ID class enum (uint8_t) has roughly made up for it.

I didn't necessarily want to use enum ORB_PRIO everywhere, but I was a bit appalled when I saw the error in sf1xx. Longer term we might be able to make it an enum class or quite honestly the entire ORB priority concept might not really make sense to keep in this form.

File: src/modules/sensors/voted_sensors_update.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/sensors/voted_sensors_update.cpp:712]: Yes I suppose so. Unrelated to this PR I came across a case where this was stuck spamming continuously on a failure, so I was actually planning to take a closer look soon.

Comment (dagar) [inline src/modules/sensors/voted_sensors_update.cpp:712]:
```suggestion
					if (sensor.priority[i] > ORB_PRIO_MIN) {
```

**Comment by dagar** [comment]
TODO: manually verify queued publication unadavertise behavior.

**Comment by dagar** [comment]
> We need to double-check if that works as expected. No case specifically comes to mind, ulog_stream_ack might be a candidate.

I've stepped through this a few times and verified we're unadvertising regular Publications and leaving Queued.

**Comment by dagar** [comment]
> Can you test that log streaming still works by restarting it several times?

Seems to be working.

``` Console

nsh> logger status
INFO  [logger] Running in mode: all
INFO  [logger] Number of subscriptions: 120 (3840 bytes)
INFO  [logger] Full File Logging Running:
INFO  [logger] Log file: /fs/microsd/log/sess006/log001.ulg
INFO  [logger] Wrote 1.80 MiB (avg 14.51 KiB/s)
INFO  [logger] Since last status: dropouts: 261 (max len: 0.677 s), max used buffer: 13009 / 14336 B


nsh> INFO  [logger] Start mavlink log


nsh>  logger status
INFO  [logger] Running in mode: all
INFO  [logger] Number of subscriptions: 120 (3840 bytes)
INFO  [logger] Full File Logging Running:
INFO  [logger] Log file: /fs/microsd/log/sess006/log001.ulg
INFO  [logger] Wrote 2.70 MiB (avg 14.32 KiB/s)
INFO  [logger] Since last status: dropouts: 1 (max len: 0.517 s), max used buffer: 12723 / 14336 B
INFO  [logger] Mavlink Logging Running (Full log)


nsh> mavlink status

instance #0:
        GCS heartbeat:  269194 us ago
        mavlink chan: #0
        type:           USB CDC
        flow control: OFF
        rates:
          tx: 79.366 kB/s
          txerr: 1.515 kB/s
          tx rate mult: 0.981
          tx rate max: 800000 B/s
          rx: 0.598 kB/s
        ULog rate: 6.7% of max 70.0%
        FTP enabled: YES, TX enabled: YES
        mode: Config
        MAVLink version: 2
        transport protocol: serial (/dev/ttyACM0 @2000000)
        ping statistics:
          last: 6.84 ms
          mean: 7.79 ms
          max: 107.71 ms
          min: 0.27 ms
          dropped packets: 0


nsh> INFO  [logger] Stop mavlink log
nsh> INFO  [logger] Start mavlink log
nsh> logger status
INFO  [logger] Running in mode: all
INFO  [logger] Number of subscriptions: 120 (3840 bytes)
INFO  [logger] Full File Logging Running:
INFO  [logger] Log file: /fs/microsd/log/sess006/log001.ulg
INFO  [logger] Wrote 3.65 MiB (avg 13.85 KiB/s)
INFO  [logger] Since last status: dropouts: 141 (max len: 0.843 s), max used buffer: 0 / 14336 B
INFO  [logger] Mavlink Logging Running (Full log)

nsh> INFO  [logger] Stop mavlink log

nsh> logger status
INFO  [logger] Running in mode: all
INFO  [logger] Number of subscriptions: 120 (3840 bytes)
INFO  [logger] Full File Logging Running:
INFO  [logger] Log file: /fs/microsd/log/sess006/log001.ulg
INFO  [logger] Wrote 3.87 MiB (avg 13.85 KiB/s)
INFO  [logger] Since last status: dropouts: 51 (max len: 0.704 s), max used buffer: 0 / 14336 B

nsh> INFO  [logger] Start mavlink log
nsh> INFO  [logger] Stop mavlink log
```

https://review.px4.io/plot_app?log=512007ef-4fee-4cd9-8594-d449cccad7f2
https://review.px4.io/plot_app?log=c9799a7d-fe83-475e-95fb-2f8fa78185b1
https://review.px4.io/plot_app?log=469c70b1-8ea4-4d4a-b7b9-857e6338fa62

Side question, is it possible to capture postflight additional data when log streaming is stopped?

**Comment by bkueng** [comment]
> Side question, is it possible to capture postflight additional data when log streaming is stopped?

Not easily, as stopping is requested externally.

**Pull request merged**